Feature/device specific test minimo 400

enicky opened the pull request:
Added a complete example for my Gardena Minimo 400.

I extended the mower.py with the extra needed commands. There were also a few small bugs that appeared while testing those commands on my mower.

erikmol commented on automower_ble/protocol.json:
```diff
@@ -184,7 +184,8 @@
         "requestType":{
-            "task": "uint8"
+            "unknown": "uint16",
+            "task": "uint16"
```
```suggestion
            "task": "uint32"
```
According to Automower apk this is a uint32. As you append `00 00` by setting `unknown` this code still works as you're essentially converting a uint16.

erikmol commented on device_tests/test_minimi_400.py:
```diff
@@ -0,0 +1,210 @@
+
+            logger.debug('--------------')
+            logger.debug('start setting mode to manual')
+            await mower.set_mode_of_operation("manual")
+            logger.debug('Mode of operation set to manual')
+            logger.debug('--------------')
+
+
+            logger.debug(f'Overriding mow to 30 mins')
+            await mower.set_mower_override_duration_in_seconds(30*60) # 30 minutes override mow
+            logger.debug(f'override mow finished ')
+            logger.debug('--------------')
+
+
+            start_trigger = await mower.start_trigger_request()
+            logger.debug(f'Start trigger response : {start_trigger}')
+            
+        if must_stop:
+            logger.debug(f'Must stop mowing. Send Park command to mower')
+            await mower.set_mode_of_operation('manual')
```
`mower.set_mode_of_operation` takes an `ModeOfOperation` enum, not `str`.

enicky replied:
changed my code according to comment

erikmol commented on automower_ble/mower.py:
```diff
@@ -111,23 +116,68 @@ async def mower_resume(self):
+            return None
+        return ModeOfOperation(mode_response)
+    
+    async def set_mode_of_operation(self, mode:ModeOfOperation):
+        await self.set_parameter('setModeOfOperation', mode=mode )
```
```suggestion
        await self.set_parameter('setModeOfOperation', mode=mode.value )
```
The value of the Enum should be passed in. I now notice this should also be changed in `mower.mower_override` function.

enicky replied:
Updated code based on comments

alistair23 commented on automower_ble/mower.py:
```diff
@@ -38,6 +38,7 @@ async def get_parameter(self, parameter_name: str, **kwargs):
         command = Command(self.channel_id, self.protocol[parameter_name])
         request = command.generate_request(**kwargs)
         response = await self._request_response(request)
+        logger.debug(f'[get_parameter] response {response}')
```
We don't need to add a debug print here

enicky replied:
removed code

alistair23 commented on device_tests/test_minimi_400.py:
```diff
@@ -0,0 +1,212 @@
+import argparse
```
Can you document what this script is for

enicky replied:
added documentation in the script itself

enicky commented:
Hey @erikmol , I'm still working on it. But I had a few days off (vacation in Belgium). So in the next few days I hope to complete this

enicky commented:
Fixed the merge issues and the comments. Also formatting ...

erikmol commented:
In my point of view, it would be better to update `mower.py` with additions rather than having a dedicated `device_test_minimi_400.py`. As far as we know there should not be any differences in the communication protocol between the various brands?

There is however some general confusion about what the various commands do, in what order they should be sent, which command is valid according to the state of the mower and so on. For example:

- #33 introduced `requestTrigger`, but this command was already available as `resume`.
- What does `keepalive` mean and when do we need to send it?
- The `override` command have `no_response`, but according to a decompiled apk it has (see [this commit](https://github.com/alistair23/AutoMower-BLE/pull/50/commits/5bd5405e3010b4b0a5e4b1649d348ceb703e6e07) in my draft PR)

enicky commented:
> In my point of view, it would be better to update mower.py with additions rather than having a dedicated device_test_minimi_400.py. As far as we know there should not be any differences in the communication protocol between the various brands?

Correct. But the problem I had in the beginnen was that no command was being accepted by my mower. So I was wondering if there was a difference between the models. . So I started to check what was being sent from my phone to the mower. And I didn't want to mess up the existing mower.py ... That's why I introduced this "test". You can see it as an example of how to handle the code ... 
You are right with "there shouldn't be any differences" but again: my mower wouldn't start or stop mowing using the example mower.py. So I implemented the commands that were being sent to my mower, in the same order as they are being sent by the app. 

> What does keepalive mean and when do we need to send it?
I guess it's like a heartbeat. I see it now and then coming through the communication between the mower and the app. Maybe it's something the app sends for not letting the mower getting into a sleep - mode (getting disconnected from the app)

But, @alistair23  asked me to split up my initial PR, so that's why I ended up with these changes. If you don't need them. That's ok for me. You can cherry pick the bug - fix and reject this PR and I'll keep this at my own branch for testing purposes.

alistair23 commented:
I agree about the `device_test_minimi_400.py` file. I think we should aim for the default setup (which is then used by HA) to work for everyone. If something is missing in there we can add it to the default setup so then it benefits everyone.

That's an interesting point about the heartbeat. HA queries the mower every minute, so I'm not sure if we need it.

Overall I think the other parts of the PR are all fine. It ended up being one big PR again instead of split up into smaller segments. It probably makes sense to drop the `device_test_minimi_400.py` changes from this PR. Then work on getting your mower working by default

larsvinc commented:
Tried to run the test on my minimo today.  It worked except start and stop commands: 
2024-06-11 17:23:44,541 __main__ WARNING: Must start mowing
2024-06-11 17:23:44,553 asyncio  DEBUG: Using selector: EpollSelector
2024-06-11 17:23:44,554 __main__ INFO: Start test mower
2024-06-11 17:23:44,729 __main__ INFO: Device found. Start connecting to device 74:46:B3:39:CB:B2
2024-06-11 17:23:52,942 __main__ INFO: Connected to device with address 74:46:B3:39:CB:B2
2024-06-11 17:23:52,943 __main__ INFO: Mower : <automower_ble.mower.Mower object at 0x7fa5ffacd0>
2024-06-11 17:23:53,039 __main__ INFO: Model : SILENO Minimo Unknown (29,7)
2024-06-11 17:23:53,040 __main__ INFO: Start sending keep Alive
2024-06-11 17:23:54,843 automower_ble.mower ERROR: Response failed validation
2024-06-11 17:23:54,844 __main__ DEBUG: Enter Operator Pin Request Result True
2024-06-11 17:23:54,942 __main__ DEBUG: Startupsequence is required response 0
2024-06-11 17:23:55,038 __main__ DEBUG: Operator is logged in 0
2024-06-11 17:23:55,137 __main__ DEBUG: Mower activity : MowerActivity.NONE
2024-06-11 17:23:55,234 __main__ DEBUG: Mower mode : ModeOfOperation.AUTO
2024-06-11 17:23:55,380 __main__ DEBUG: Serial number : 230676374
2024-06-11 17:23:55,478 __main__ DEBUG: Restriction Reason : RestrictionReason.NONE
2024-06-11 17:23:55,576 __main__ DEBUG: No next start time
2024-06-11 17:23:55,720 __main__ DEBUG: number of tasks : 1
2024-06-11 17:23:55,720 __main__ DEBUG: Start requesting task 0
2024-06-11 17:23:55,818 __main__ DEBUG: Next task starts at 04:00:00
2024-06-11 17:23:55,819 __main__ DEBUG: Next task has a duration of 4500 seconds
2024-06-11 17:23:55,819 __main__ DEBUG: Next task must run on MONDAY    True
2024-06-11 17:23:55,819 __main__ DEBUG: Next task must run on TUESDAY   True
2024-06-11 17:23:55,820 __main__ DEBUG: Next task must run on WEDNESDAY True
2024-06-11 17:23:55,820 __main__ DEBUG: Next task must run on THURSDAT  True
2024-06-11 17:23:55,820 __main__ DEBUG: Next task must run on FRIDAY    True
2024-06-11 17:23:55,820 __main__ DEBUG: Next task must run on SATURDAY  True
2024-06-11 17:23:55,821 __main__ DEBUG: Next task must run on SUNDAY    True
2024-06-11 17:23:56,014 __main__ DEBUG: Mower state response MowerState.STOPPED
2024-06-11 17:23:56,111 __main__ DEBUG: Mower activity : MowerActivity.NONE
2024-06-11 17:23:56,208 __main__ DEBUG: Mode : ModeOfOperation.AUTO
2024-06-11 17:23:56,355 __main__ DEBUG: Next Start Time : None
2024-06-11 17:23:56,355 __main__ DEBUG: No next start time
2024-06-11 17:23:56,453 __main__ DEBUG: Restriction reason : RestrictionReason.NONE
2024-06-11 17:23:56,453 __main__ DEBUG: --------------
2024-06-11 17:23:56,454 __main__ DEBUG: start setting mode to manual
2024-06-11 17:23:56,549 automower_ble.mower ERROR: Response failed validation
2024-06-11 17:23:56,550 __main__ DEBUG: Mode of operation set to manual
2024-06-11 17:23:56,550 __main__ DEBUG: --------------
2024-06-11 17:23:56,551 __main__ DEBUG: Overriding mow to 30 mins
2024-06-11 17:23:56,696 automower_ble.mower ERROR: Response failed validation
2024-06-11 17:23:56,696 __main__ DEBUG: override mow finished
2024-06-11 17:23:56,697 __main__ DEBUG: --------------
2024-06-11 17:23:56,794 automower_ble.mower ERROR: Response failed validation
2024-06-11 17:23:56,795 __main__ DEBUG: Start trigger response : None
2024-06-11 17:23:56,988 __main__ DEBUG: Mower state response MowerState.STOPPED
2024-06-11 17:23:57,086 __main__ DEBUG: Mower activity : MowerActivity.NONE
2024-06-11 17:23:57,086 __main__ INFO: Finished testing mower
2024-06-11 17:23:59,066 __main__ INFO: Disconnected from Mower

I had to connect during the 3 min paring window, and I had to use the pin code in order to connect.

Let me know how I can help developing this library.
